src/zope/testrunner/runner.py

```diff
@@ -317,11 +318,22 @@ def run_tests(self):
         if setup_layers:
             if self.options.resume_layer is None:
                 self.options.output.info("Tearing down left over layers:")
-            tear_down_unneeded(self.options, (), setup_layers, True)
+            tear_down_unneeded(
+                self.options, (), setup_layers, self.errors, optional=True)
 
         self.failed = bool(self.import_errors or self.failures or self.errors)
 
 
+def handle_layer_failure(failure_type, output, errors):
+    if hasattr(output, 'layer_failure'):
+        output.layer_failure(failure_type.subunit_label, sys.exc_info())
+    else:
+        f = StringIO()
+        traceback.print_exc(file=f)
+        output.error(f.getvalue())
```

Comment:
Why the StringIO() dance rather than `traceback.format_exc()`?  (I've a feeling there's a good reason.)

Reply:
Ah, you moved preexisting code here, so I may be asking the wrong person.